gpt_index/vector_stores/metal.py

```diff
@@ -0,0 +1,116 @@
+import math
+from typing import Any, Dict, List, Optional
+from gpt_index.data_structs.node_v2 import Node
+from gpt_index.vector_stores.types import VectorStore, NodeEmbeddingResult, VectorStoreQuery, VectorStoreQueryResult
+
+
+class MetalVectorStore(VectorStore):
+    def __init__(self, api_key: str, client_id: str, index_id: str, filters: Optional[Dict[str, Any]] = None):
+        """Init params."""
+        import_err_msg = (
+            "`metal_sdk` package not found, please run `pip install metal_sdk`"
+        )
+        try:
+            import metal_sdk  # noqa: F401
+        except ImportError:
+            raise ImportError(import_err_msg)
+        from metal_sdk.metal import Metal   # noqa: F401
+
+
+        self.api_key = api_key
+        self.client_id = client_id
+        self.index_id = index_id
+        self.filters = filters
+
+        self.metal_client = Metal(api_key, client_id, index_id)
+        self.stores_text = True
+        self.is_embedding_query = True
+
+    @classmethod
+    def from_dict(cls, config_dict: Dict[str, Any]) -> "MetalVectorStore":
+        return cls(
+            api_key=config_dict["api_key"],
+            client_id=config_dict["client_id"],
+            index_id=config_dict["index_id"],
+            filters=config_dict.get("filters", None)
+        )
+
+    def query(self, query: VectorStoreQuery) -> VectorStoreQueryResult:
+        payload = {
+            "text": query.query_str,  # Query Text
+            "filters": self.filters,  # Metadata Filters
+        }
```

Comment:
To clarify: metal embeds text into dense vectors under the hood?

The reason I'm asking is: in LLamaIndex, we allow user to customize the embedding model. 
And right now we always call the embedding model AFAIK. 

I think currently, that computation would be duplicated (once in LlamaIndex and not used) and then in Metal.

Reply:
ohhh i understand, so the expectation is that embedding gen happens before it hits the vectorstore abstraction? just updated to accept the embedding into our api